external-crates/move/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -140,84 +126,706 @@ struct PackageWithResolverTOML<'a>(&'a Package);
 struct DependencyTOML<'a>(PM::PackageName, &'a Dependency);
 struct SubstTOML<'a>(&'a PM::Substitution);
 
-impl DependencyGraph {
+/// A builder for `DependencyGraph`
+pub struct DependencyGraphBuilder<Progress: Write> {
+    /// Used to avoid re-fetching dependencies that are already present locally
+    pub dependency_cache: DependencyCache,
+    /// Logger
+    pub progress_output: Progress,
+    /// A chain of visited dependencies used for cycle detection
+    visited_dependencies: VecDeque<(PM::PackageName, PM::InternalDependency)>,
+}
+
+impl<Progress: Write> DependencyGraphBuilder<Progress> {
+    pub fn new(skip_fetch_latest_git_deps: bool, progress_output: Progress) -> Self {
+        DependencyGraphBuilder {
+            dependency_cache: DependencyCache::new(skip_fetch_latest_git_deps),
+            progress_output,
+            visited_dependencies: VecDeque::new(),
+        }
+    }
+
+    /// Get a graph from the Move.lock file, if Move.lock file is present and up-to-date
+    /// (additionally returning false), otherwise compute a new graph based on the content of the
+    /// Move.toml (manifest) file (additionally returning true).
+    pub fn get_graph(
+        &mut self,
+        parent: &PM::DependencyKind,
+        root_path: PathBuf,
+        manifest_string: String,
+        lock_string: Option<String>,
+    ) -> Result<(DependencyGraph, bool)> {
+        let toml_manifest = parse_move_manifest_string(manifest_string.clone())?;
+        let manifest = parse_source_manifest(toml_manifest)?;
+
+        // compute digests eagerly as even if we can't reuse existing lock file, they need to become
+        // part of the newly computed dependency graph
+        let new_manifest_digest = digest_str(manifest_string.into_bytes().as_slice());
+
+        let new_deps_digest_opt = self.dependency_digest(root_path.clone(), &manifest)?;
```

Comment:
Makes sense but I as it would take as much time to do it now as it would in the following PR, I'd indeed prefer to do this change separately. I can work on this in the PR immediately following this one.